File: ui/components/ui/ProducerSession/message-generator/deliver-at-generator/deliver-at-generator.ts
import * as pb from '../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { DateTimeGenerator, dateTimeGeneratorFromPb, dateTimeGeneratorToPb } from '../data-generators/date-time/date-time-generator';

export type DeliverAtGenerator = {
  type: 'deliver-at-generator',
  generator: DateTimeGenerator
};

export function deliverAtGeneratorFromPb(v: pb.DeliverAtGenerator): DeliverAtGenerator {
  return {
    type: 'deliver-at-generator',
    generator: dateTimeGeneratorFromPb(v.getGenerator()!)
  };
}

export function deliverAtGeneratorToPb(v: DeliverAtGenerator): pb.DeliverAtGenerator {
  const generatorPb = new pb.DeliverAtGenerator();
  generatorPb.setGenerator(dateTimeGeneratorToPb(v.generator));

  return generatorPb;
}

File: server/src/main/scala/producer/message_generator/deliver_at_generator/DeliverAtGenerator.scala
package producer.message_generator.deliver_at_generator

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import producer.message_generator.data_generators.date_time.DateTimeGenerator

case class DeliverAtGenerator(
    generator: DateTimeGenerator
)

object DeliverAtGenerator:
    def fromPb(v: pb.DeliverAtGenerator): DeliverAtGenerator =
        DeliverAtGenerator(
            generator = DateTimeGenerator.fromPb(v.generator.get)
        )

    def toPb(v: DeliverAtGenerator): pb.DeliverAtGenerator =
        pb.DeliverAtGenerator(
            generator = Some(DateTimeGenerator.toPb(v.generator))
        )
